### PR TITLE
Mixed distillation loss

### PR DESCRIPTION
**Pull request by jlamypoirier**
# ✨ Description

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [x] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

### CODE REVIEW

**Comment by oleksost** [comment]
@jlamypoirier when you have time to have a quick look

File: fast_llm/functional/cross_entropy.py [hunk 1]
```diff
@@ -28,6 +27,8 @@ def _torch_cross_entropy_forward_backward(
         if target_format == TargetFormat.logits:
             if logits_scale_factor != 1.0:
                 target = target * logits_scale_factor
+            if teacher_softmax_temp != 1.0:
+                target = target * (1 / teacher_softmax_temp)
```
Comment (jlamypoirier) [inline fast_llm/functional/cross_entropy.py:31]: `target/teacher_softmax_temp `?

File: fast_llm/functional/cross_entropy.py [hunk 2]
```diff
@@ -14,6 +12,7 @@ def _torch_cross_entropy_forward_backward(
     loss_mask: torch.Tensor | None,
     grad_output: float | None,
     logits_scale_factor: float,
+    teacher_softmax_temp: float,
```
Comment (jlamypoirier) [inline fast_llm/functional/cross_entropy.py:15]: `teacher_softmax_temperature`

Comment (jlamypoirier) [inline fast_llm/functional/cross_entropy.py:15]: `= 1.0`

File: fast_llm/functional/cross_entropy.py [hunk 3]
```diff
@@ -95,7 +97,7 @@ def _fused_cross_entropy_forward_backward(
     logits_norm, exp_logits, sum_exp_logits = _fused_softmax_base(logits, logits_scale_factor, group)
 
     if target_format == TargetFormat.logits:
-        target = _fused_softmax(target, logits_scale_factor, group)
+        target = _fused_softmax(target, logits_scale_factor * (1 / teacher_softmax_temp), group)
```
Comment (jlamypoirier) [inline fast_llm/functional/cross_entropy.py:100]: `/`?

File: fast_llm/functional/cross_entropy.py [hunk 4]
```diff
@@ -189,9 +192,137 @@ def cross_entropy_forward_backward(
...
+            logits, target, loss_mask, grad_output, logits_scale_factor, teacher_softmax_temp, target_format, group
         )
     else:
         return _CROSS_ENTROPY_IMPLEMENTATIONS[implementation](
-            logits, target, loss_mask, grad_output, logits_scale_factor, target_format
+            logits, target, loss_mask, grad_output, logits_scale_factor, teacher_softmax_temp, target_format
         )
+
+
+def _torch_reverse_kl_forward_backward(
```
Comment (jlamypoirier) [inline fast_llm/functional/cross_entropy.py:203]: I recommend adding test(s). Also why separating from `reverse_kl_forward_backward`?

File: fast_llm/layers/language_model/config.py [hunk 5]
```diff
@@ -111,6 +114,11 @@ class LanguageModelBaseConfig(BaseModelConfig):
         desc="Implementation for the cross-entropy computation.",
         hint=FieldHint.performance,
     )
+    distil_loss_impl: CrossEntropyImpl = Field(
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/config.py:117]: `distillation_loss_implementation`

File: fast_llm/layers/language_model/config.py [hunk 6]
```diff
@@ -149,6 +157,16 @@ class LanguageModelBaseConfig(BaseModelConfig):
         hint=FieldHint.feature,
         valid=check_field(Assert.geq, 0),
     )
+    distil_ntp_loss_factor: float = Field(
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/config.py:160]: ntp?

File: fast_llm/layers/language_model/config.py [hunk 7]
```diff
@@ -149,6 +157,16 @@ class LanguageModelBaseConfig(BaseModelConfig):
...
+        default=0.0,
+        desc="Factor to scale the distillation loss by.",
+        hint=FieldHint.feature,
+    )
+    distil_loss_factor: float = Field(
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/config.py:165]: distillation_loss_factor

Comment (jlamypoirier) [reply fast_llm/layers/language_model/config.py:165]: Also it's the same as `distil_ntp_loss_factor` according to the description?

File: fast_llm/layers/language_model/config.py [hunk 8]
```diff
@@ -157,6 +175,13 @@ class LanguageModelBaseConfig(BaseModelConfig):
         hint=FieldHint.feature,
         valid=check_field(Assert.geq, 0),
     )
+    teacher_softmax_temp: float = Field(
+        default=1.0,
+        desc="Multiplies target logits by 1/temperature inc ase of distillation.",
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/config.py:180]: Divides distillation target logits by this factor.

File: fast_llm/layers/language_model/head.py [hunk 9]
```diff
@@ -205,7 +211,9 @@ def _get_targets(
                 if loss_mask is not None:
                     loss_mask = loss_mask.flatten()
 
-            if self._config.distillation_model is None:
+            if self._config.distillation_model is None or (
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/head.py:214]: `self._config.distillation_model is not None` is redundant

File: fast_llm/layers/language_model/head.py [hunk 10]
```diff
@@ -369,30 +377,51 @@ def _logits_cross_entropy_forward_backward(
             dpo_loss, dpo_grad = None, None
 
         if lm_target is not None:
+            distil_ntp_loss_factor = 1.0
+            if self._config.distillation_model is not None:
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/head.py:381]: `distil_ntp_loss_factor = 1.0 if self._config.distillation_model is None else distil_ntp_loss_factor `

Comment (jlamypoirier) [reply fast_llm/layers/language_model/head.py:381]: Not sure that's the best option, lm loss has nothing to do with distillation so I recommend:
* Use variable `language_model_loss_factor : float = None`
* Set implicit default in validation to 1.0 or 0.0
* Always use this factor here.

File: fast_llm/layers/language_model/head.py [hunk 11]
```diff
@@ -369,30 +377,51 @@ def _logits_cross_entropy_forward_backward(
...
                 lm_target,
                 None,
                 group=self._tensor_space.distributed.tensor_group if self._parallel_embeddings else None,
-                grad_output=grad_output * self._loss_coefficient,
+                grad_output=grad_output * self._loss_coefficient * distil_ntp_loss_factor,
                 implementation=self._cross_entropy_impl,
                 logits_scale_factor=self._logits_scale_factor,
                 target_format=TargetFormat.labels,
             )
+            lm_loss = lm_loss * distil_ntp_loss_factor
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/head.py:395]: Skip if factor is 1

File: fast_llm/functional/config.py [hunk 12]
```diff
@@ -94,6 +94,7 @@ class CrossEntropyImpl(str, enum.Enum):
     torch = "torch"
     fused = "fused"
     triton = "triton"
+    reverse_kl = "reverse_kl"
```
Comment (jlamypoirier) [inline fast_llm/functional/config.py:97]: This isn't a different implementation, it's a different loss altogether.

File: fast_llm/layers/language_model/head.py [hunk 13]
```diff
@@ -369,30 +377,51 @@ def _logits_cross_entropy_forward_backward(
...
+                    ),
+                )
+            else:
+                distillation_loss, distillation_grad = cross_entropy_forward_backward(
+                    logits.flatten(0, -2),
+                    distillation_target,
+                    loss_mask,
+                    group=self._tensor_space.distributed.tensor_group if self._parallel_embeddings else None,
+                    grad_output=grad_output * self._loss_coefficient * self._distil_loss_factor,
+                    implementation=self._cross_entropy_impl,
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/head.py:420]: `_distil_loss_impl`?

File: fast_llm/layers/language_model/config.py [hunk 14]
```diff
@@ -22,6 +22,9 @@ class LanguageModelDimNames:
 class LanguageModelLossNames:
     language_model_loss = "language_model_loss"
     z_loss = "z_loss"
+    dpo_loss = "dpo_loss"
+    distil_lm_loss = "distil_lm_loss"
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/config.py:26]: Isn't this the same as `language_model_loss`?

File: fast_llm/functional/cross_entropy.py (hunk 4 above)
Comment (oleksost) [reply fast_llm/functional/cross_entropy.py:203]: Will add a test. 

`Also why separating from reverse_kl_forward_backward?`

Ideally, we would also have a fuser version of reverse kl later.

**Comment by tscholak** [comment]
It's because @jlamypoirier opened the PR 😉

File: fast_llm/layers/language_model/head.py [hunk 15]
```diff
@@ -336,30 +357,92 @@ def _logits_cross_entropy_forward_backward(
...
+
+        # TODO: Return individual losses?
+        loss = _add_tensors(dpo_loss, lm_loss, distillation_loss)
+        if self.training and losses is not None:
+            if dpo_loss is not None:
+                losses[LanguageModelLossNames.dpo_loss].append(dpo_loss.detach())
+            if distillation_loss is not None:
+                losses[LanguageModelLossNames.distillation_loss].append(distillation_loss.detach())
+            if self._config.distillation_model is not None and lm_loss is not None:
+                losses[LanguageModelLossNames.distil_lm_loss].append(lm_loss.detach())
```
Comment (tscholak) [inline fast_llm/layers/language_model/head.py:436]: does this need a new label or can this just be lm loss?

File: fast_llm/layers/language_model/head.py [hunk 16]
```diff
@@ -336,30 +357,92 @@ def _logits_cross_entropy_forward_backward(
...
+                None,
                 group=self._tensor_space.distributed.tensor_group if self._parallel_embeddings else None,
-                grad_output=grad_output,
+                grad_output=grad_output * self._loss_coefficient * self._language_model_loss_factor,
                 implementation=self._cross_entropy_impl,
                 logits_scale_factor=self._logits_scale_factor,
-                target_format=TargetFormat.labels if self._config.distillation_model is None else TargetFormat.logits,
+                target_format=TargetFormat.labels,
             )
+            lm_loss = lm_loss * self._language_model_loss_factor
```
Comment (tscholak) [inline fast_llm/layers/language_model/head.py:388]: is that factor 1.0 if unused?

**Review by tscholak — APPROVED** [review]
LGTM! thanks!

File: fast_llm/layers/language_model/head.py (hunk 15 above)
Comment (tscholak) [reply fast_llm/layers/language_model/head.py:436]: > language_model_loss is the final combined loss

ah, I see! Ok, that isn't very clear, but I guess why this is the case for historical reasons.

**Comment by oleksost** [comment]
@jlamypoirier I think I addressed all (most of) your comments (test, CrossEntropyLossImp, etc.) . Can this be merged?
